File: app/src/main/java/ch/threema/app/routines/UpdateAppLogoRoutine.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2016-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.routines;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.slf4j.Logger;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.Calendar;
import java.util.Date;

import javax.net.ssl.HttpsURLConnection;

import ch.threema.app.services.FileService;
import ch.threema.app.services.PreferenceService;
import ch.threema.app.utils.ConfigUtils;
import ch.threema.app.utils.FileUtil;
import ch.threema.app.utils.TestUtil;
import ch.threema.base.utils.LoggingUtil;

import static android.provider.MediaStore.MEDIA_IGNORE_FILENAME;

/**
 * Update the app icon
 */
public class UpdateAppLogoRoutine implements Runnable {
    private static final Logger logger = LoggingUtil.getThreemaLogger("UpdateAppLogoRoutine");

    private final FileService fileService;
    private final PreferenceService preferenceService;
    private final String lightUrl;
    private final String darkUrl;
    private boolean running = false;
    private final boolean forceUpdate;

    public UpdateAppLogoRoutine(FileService fileService,
                                PreferenceService preferenceService,
                                @Nullable String lightUrl,
                                @Nullable String darkUrl,
                                boolean forceUpdate) {
        this.fileService = fileService;
        this.preferenceService = preferenceService;
        this.lightUrl = lightUrl;
        this.darkUrl = darkUrl;
        this.forceUpdate = forceUpdate;
    }

    @Override
    public void run() {
        logger.debug("start update app logo {}, {}", this.lightUrl, this.darkUrl);
        this.running = true;

        //validate instances
        if (!TestUtil.required(this.fileService, this.preferenceService)) {
            this.running = false;
            logger.error("Not all required instances defined");
            return;
        }

        this.updateLogo(this.lightUrl, ConfigUtils.THEME_LIGHT);
        this.updateLogo(this.darkUrl, ConfigUtils.THEME_DARK);
        this.running = false;
    }

    private void setLogo(@NonNull String url, @NonNull File file, @NonNull Date expires, @ConfigUtils.AppThemeSetting String theme) {
        this.fileService.saveAppLogo(file, theme);
        this.preferenceService.setAppLogo(url, theme);
        this.preferenceService.setAppLogoExpiresAt(expires, theme);
    }

    private void clearLogo(@ConfigUtils.AppThemeSetting String theme) {
        logger.info("Clearing app logo for (forcedUpdate={}, theme={})", forceUpdate, theme);
        this.fileService.saveAppLogo(null, theme);
        this.preferenceService.clearAppLogo(theme);
    }

    private void updateLogo(@Nullable String urlString, @ConfigUtils.AppThemeSetting String theme) {
        logger.info("Update app logo (forcedUpdate={}, theme={})", forceUpdate, theme);
        Date now = new Date();
        //get expires date

        if (TestUtil.isEmptyOrNull(urlString)) {
            this.clearLogo(theme);
            return;
        }
        // Check expiry date only if update is not forced
        if (!this.forceUpdate) {
            Date expiresAt = this.preferenceService.getAppLogoExpiresAt(theme);

            if (expiresAt != null && now.before(expiresAt)) {
                logger.info("Logo not expired");
                //do nothing!
                return;
            }
        }

        //define default expiry date (now + 1day)
        Calendar tomorrowCalendar = Calendar.getInstance();
        tomorrowCalendar.setTime(now);
        tomorrowCalendar.add(Calendar.DATE, 1);
        Date tomorrow = tomorrowCalendar.getTime();

        try {
            logger.info("Download {}", urlString);

            URL url = new URL(urlString);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setSSLSocketFactory(ConfigUtils.getSSLSocketFactory(url.getHost()));

            try {
                // Warning: This may implicitly open an error stream in the 4xx/5xx case!
                connection.connect();
                final int responseCode = connection.getResponseCode();
                if (responseCode != HttpsURLConnection.HTTP_OK) {
                    if (responseCode == HttpsURLConnection.HTTP_NOT_FOUND) {
                        logger.warn("Logo not found");
                    } else {
                        logger.warn("Connection failed with response code {}", responseCode);
                    }
                } else {
                    logger.debug("Logo found. Start download");

                    File temporaryFile = this.fileService.createTempFile(MEDIA_IGNORE_FILENAME, "appicon");
                    // this will be useful to display download percentage
                    // might be -1: server did not report the length
                    int fileLength = connection.getContentLength();
                    logger.debug("size: {}", fileLength);

                    // download the file
                    try (InputStream input = connection.getInputStream()) {
                        Date expires = new Date(connection.getHeaderFieldDate("Expires", tomorrow.getTime()));
                        logger.debug("expires {}", expires);

                        try (FileOutputStream output = new FileOutputStream(temporaryFile.getPath())) {
                            byte[] data = new byte[4096];
                            int count;

                            while ((count = input.read(data)) != -1) {
                                output.write(data, 0, count);
                            }

                            logger.info("Logo downloaded. Expires at {}.", expires);
                            output.close();

                            this.setLogo(urlString, temporaryFile, expires, theme);

                            FileUtil.deleteFileOrWarn(temporaryFile, "temporary file", logger);

                        } catch (IOException x) {
                            // Failed to download
                            // do nothing an try again later
                            logger.error("Download of app logo failed", x);
                        }
                    }
                }
            } finally {
                try {
                    final InputStream errorStream = connection.getErrorStream();
                    if (errorStream != null) {
                        errorStream.close();
                    }
                } catch (IOException e) {
                    // Ignored
                }
            }
        } catch (Exception x) {
            logger.error("Update of app logo failed", x);
        }

    }

    protected boolean isRunning() {
        return this.running;
    }
}

File: app/src/main/java/ch/threema/app/routines/UpdateAppLogoRoutine.kt
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2016-2025 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.routines

import android.provider.MediaStore
import ch.threema.app.preference.service.PreferenceService
import ch.threema.app.services.FileService
import ch.threema.app.utils.ConfigUtils
import ch.threema.app.utils.ConfigUtils.AppThemeSetting
import ch.threema.base.utils.getThreemaLogger
import ch.threema.common.Http
import ch.threema.common.TimeProvider
import ch.threema.common.buildRequest
import ch.threema.common.copyIntoFile
import ch.threema.common.execute
import ch.threema.common.getExpiration
import ch.threema.common.getSuccessBodyOrThrow
import ch.threema.common.plus
import java.io.File
import java.io.IOException
import java.time.Instant
import kotlin.time.Duration.Companion.days
import okhttp3.OkHttpClient

private val logger = getThreemaLogger("UpdateAppLogoRoutine")

class UpdateAppLogoRoutine(
    private val fileService: FileService,
    private val preferenceService: PreferenceService,
    private val okHttpClient: OkHttpClient,
    private val lightUrl: String?,
    private val darkUrl: String?,
    private val forceUpdate: Boolean,
    private val timeProvider: TimeProvider = TimeProvider.default,
) : Runnable {
    override fun run() {
        logger.debug("start update app logo {}, {}", lightUrl, darkUrl)
        updateLogo(lightUrl, ConfigUtils.THEME_LIGHT)
        updateLogo(darkUrl, ConfigUtils.THEME_DARK)
    }

    private fun updateLogo(logoUrl: String?, @AppThemeSetting theme: String) {
        logger.info("Update app logo (forcedUpdate={}, theme={})", forceUpdate, theme)
        if (logoUrl.isNullOrEmpty()) {
            clearLogo(theme)
            return
        }

        val now = timeProvider.get()

        if (!forceUpdate) {
            val expiresAt = preferenceService.getAppLogoExpiresAt(theme)
            if (expiresAt != null && now < expiresAt) {
                logger.info("Logo not expired")
                return
            }
        }

        try {
            logger.info("Downloading {}", logoUrl)
            val request = buildRequest {
                get()
                url(logoUrl)
            }
            okHttpClient.execute(request).use { response ->
                when {
                    response.isSuccessful -> {
                        logger.debug("Logo found. Start download")
                        val temporaryFile = fileService.createTempFile(MediaStore.MEDIA_IGNORE_FILENAME, "appicon")
                        try {
                            val expires = response.getExpiration() ?: (now + 1.days)
                            response.getSuccessBodyOrThrow()
                                .copyIntoFile(temporaryFile)
                            logger.info("Logo downloaded. Expires at {}.", expires)

                            setLogo(logoUrl, temporaryFile, expires, theme)
                        } finally {
                            temporaryFile.delete()
                        }
                    }

                    response.code == Http.StatusCode.NOT_FOUND -> logger.warn("Logo not found")
                    else -> logger.warn("Connection failed with response code {}", response.code)
                }
            }
        } catch (e: IOException) {
            logger.error("Update of app logo failed", e)
        }
    }

    private fun setLogo(url: String, file: File, expires: Instant, @AppThemeSetting theme: String) {
        fileService.saveAppLogo(file, theme)
        preferenceService.setAppLogo(url, theme)
        preferenceService.setAppLogoExpiresAt(expires, theme)
    }

    private fun clearLogo(@AppThemeSetting theme: String) {
        logger.info("Clearing app logo for (forcedUpdate={}, theme={})", forceUpdate, theme)
        fileService.saveAppLogo(null, theme)
        preferenceService.clearAppLogo(theme)
    }
}
